pydrive2/files.py

```diff
@@ -229,17 +232,43 @@ def GetContentFile(self, filename, mimetype=None, remove_bom=False):
     :type mimetype: str
     :param remove_bom: Whether to remove the byte order marking.
     :type remove_bom: bool
+    :param callback: passed two arguments: (total trasferred, file size).
+    :type param: callable
     :raises: ApiRequestError, FileNotUploadedError, FileNotDownloadableError
     """
-        if (
-            self.content is None
-            or type(self.content) is not io.BytesIO
-            or self.has_bom == remove_bom
-        ):
-            self.FetchContent(mimetype, remove_bom)
-        f = open(filename, "wb")
-        f.write(self.content.getvalue())
-        f.close()
+        files = self.auth.service.files()
+        file_id = self.metadata.get("id") or self.get("id")
+
+        def download(fd, request):
+            downloader = MediaIoBaseDownload(fd, request)
+            done = False
+            while done is False:
+                status, done = downloader.next_chunk()
+                if callback:
+                    callback(status.resumable_progress, status.total_size)
+
+        with open(filename, mode="w+b") as fd:
+            # Ideally would use files.export_media instead if
+            # metadata.get("mimeType").startswith("application/vnd.google-apps.")
+            # but that would first require a slow call to FetchMetadata()
+            try:
+                download(fd, files.get_media(fileId=file_id))
+            except errors.HttpError as error:
+                err_str = str(error).lower()
+                if "403" not in err_str or "use export" not in err_str:
+                    raise
+                mimetype = mimetype or "text/plain"
+                fd.seek(0)
```

Comment:
is it here "just in case"? :)

Reply:
No I checked and it's needed. It also wouldn't be right for the `_RemovePrefix` to do a `seek(0)` on its input stream.

```python
with open("fun", mode="wb+") as fd:
    fd.write("feffGeneric, non-exhaustive ASCII test string.")
    fd.seek(0)
    print(fd.read())
    # fd.seek(0)  # needed
    _RemovePrefix(fd, "feff")
    fd.seek(0)
    print(fd.read())
```
```
feffGeneric, non-exhaustive ASCII test string.
feffGeneric, non-exhaustive ASCII test string.
```

Reply:
Yeah, I was just curious how come it writes anything to it before raising an exception in the first place.

Reply:
ah srry I thought you were referring to a different `seek(0)`. Right; yes "just in case". I'm not too opinionated on whether or not this is too defensive programming.

Reply:
np, up to you to keep ... may put a small comment nearby, so that other people won't be looking for a hidden reason here :)